--- a/QuantumDots/src/HilbertSpace/ThreeDTwoParticles.cc
+++ b/QuantumDots/src/HilbertSpace/ThreeDTwoParticles.cc
@@ -141,6 +141,8 @@
   int dimension2 = this->SecondParticle->GetHilbertSpaceDimension ();
   int state1 = state / dimension2; int state2 = state - state1 * dimension2;
 
-  Str << "'" << this->FirstParticle->PrintState (Str, state1) << ", " << this->SecondParticle->PrintState (Str, state2) << ")";
+  Str << "(";
+  this->FirstParticle->PrintState (Str, state1) << ", ";
+  this->SecondParticle->PrintState (Str, state2) << ")";
   return Str;
 }
